### 处理入站异常

如果想处理入站异常，需要在ChannelInboundHandler 实现中重写 exceptionCaught（ChannelHandlerContext ctx,Throwable cause ） 方法。

如何响应异常，很大程度取决于应用程序。如果不实现任何处理入站异常的逻辑（或者没有消费该异常），那么netty 浆会记录该异常没被处理的事实

#### 总结

- channelHandler.exceptionCaught() 的默认实现是简单将当前异常转发给channel pipeline 的下一个 channel handler 
- 如果异常达到了channel pipeline 的尾端，他将被记录为未处理
- 要想定义自定义的处理逻辑，需要重写 exceptionCaught  方法，然后决定是否需要将异常传播出去。



### 处理出站异常

用于处理出站操作中的正常完成以及异常的选项，都基于一下的通知机制

- 每个出站操作都将返回一个ChannelFuture 注册到channelFuture 的channelFutrueListener 将在操作完成时被通知该操作是成功了还是出错了
- 几乎所有的channelOutboundHandler 上的方法都会传入一个channelPromise 的实例。作为channelFuture 的子类， channel promise 也可以被分配用于异步通知的监听器。 但是 ，channel promise 还具有提供立即通知的可写方法

setSuccess()

setFailure(Throwable cause )

添加channel future listener 只需要调用 channel future 实例上的addListener 方法，可以通过一下两种方式来实现

- 调用出站操作（如write）所返回的channelFuture 上的addListener 方法

```java
future.addListeners((ChannelFutureListener) channelFuture -> {
    if (channelFuture.isSuccess()){
        logger.info("send msg to server success ");
    }else {
        logger.error("send msg to server error as [{}]",channelFuture.cause().getMessage());
    }
});
```

- 将CHannelFutureListener添加到即将作为参数传递给ChannelOutboundHandler 的方法的channel promise 。 

```java
promise.addListener((ChannelFutureListener) channelFuture -> {
    if (channelFuture.isSuccess()){
        logger.info("write  msg to server success ");
    }else {
        logger.error("write msg to server error as [{}]",channelFuture.cause().getMessage(),channelFuture.cause());

    }
});
```